numpy/lib/shape_base.py

```diff
@@ -16,10 +16,222 @@
 __all__ = [
     'column_stack', 'row_stack', 'dstack', 'array_split', 'split',
     'hsplit', 'vsplit', 'dsplit', 'apply_over_axes', 'expand_dims',
-    'apply_along_axis', 'kron', 'tile', 'get_array_wrap'
+    'apply_along_axis', 'kron', 'tile', 'get_array_wrap', 'take_along_axis',
+    'put_along_axis'
     ]
 
 
+def _make_along_axis_idx(arr, indices, axis):
+	# compute dimensions to iterate over
+    if arr.ndim != indices.ndim:
+        raise ValueError(
+            "`indices` and `arr` must have the same number of dimensions")
+    shape_ones = (1,) * indices.ndim
+    dest_dims = list(range(axis)) + [None] + list(range(axis+1, indices.ndim))
+
+    # build a fancy index, consisting of orthogonal aranges, with the
+    # requested index inserted at the right location
+    fancy_index = []
+    for dim, n in zip(dest_dims, arr.shape):
+        if dim is None:
+            fancy_index.append(indices)
+        else:
+            ind_shape = shape_ones[:dim] + (-1,) + shape_ones[dim+1:]
+            fancy_index.append(_nx.arange(n).reshape(ind_shape))
+
+    return tuple(fancy_index)
+
+
+def take_along_axis(arr, indices, axis):
+    """
+    Take the elements described by `indices` along each 1-D slice of the given
+    `axis`, matching up subspaces of arr and indices.
+
+    This function can be used to index with the result of `argsort`, `argmax`,
+    and other `arg` functions.
+
+    This is equivalent to (but faster than) the following use of `ndindex` and
```

Comment:
Looking at this again, it might even be an idea to add
```
out_1d = out[ii + (slice(0, J),) + kk]  # corresponding view of output
for j in range(J):
    out_1d[j] = a_1d[indices_1d[j]]
```

Reply:
Note that this does at least link to `ndindex` and `s_`. I do like the extra temporaries you introduce here, but I would prefer to keep `s_`

Reply:
Yes, the `np.s_[:,]` is not as confusing as `np.s_[j,]` (though the comma looks very odd). One argument in favour of the slice as I wrote it is that it uses `M` and `J` and thus might help make clearer that those dimensions are not necessarily the same (which I think is useful). But of course one could just write `np.s_[0:M,]`, etc.

Reply:
On the flipside, using `:M` suggests that the dimension may not actually be of size M, which would also be confusing.

I've incorporated some of your ideas into the docstrings - perhaps best to create a new review thread on the updated one

Reply:
What I'd really like here is to just be able to spell it `out[*ii, j, *kk]` - I tried patching python to allow this, but wasn't able to work out where to go after patching the grammar.

Reply:
Yes, I tried the same! 

I do think your update is a lot clearer already. I would argue to keep the definition of `out = np.empty(...)` since that helps to understand what the result shape should be and makes the code already runnable. And perhaps add a comment in the example after extracting `J` along the lines of `# Note that J does not have to equal M=a.shape[axis]`